README.md
# MagicMirror² Module: Skyss
'MMM-Skyss' is a module based on the Ruter equivalent made by Cato Antonsen for displaying public transport information for the Hordaland region in Norway on a [MagicMirror²](https://magicmirror.builders/). It's using data from Skyss.no. Skyss is a registered trademark of Hordaland County Councile (Hordaland Fylkeskommune) which is not affiliated with this product. Content from Skyss APIs may be copyrighted.

![1](images/MMM-Skyss-1.png)
![2](images/MMM-Skyss-2.png) 

Current version is s1.3.0 See [changelog](CHANGELOG.md "Version history") for version history.

## Installation

Remote to your MM2-box with your terminal software and go to your MagicMirror's Module folder:
````bash
cd ~/MagicMirror/modules
````

Clone the repository:
````bash
git clone https://github.com/PabloDons/MMM-Skyss.git
````

Go to the modules folder:
````bash
cd MMM-Skyss
````

Install the dependencies:
````bash
npm install
````

Add the module to the modules array in the `config/config.js` file by adding the following section. You can change this configuration later when you see this works:
```javascript
{
	module: 'MMM-Skyss',
	header: 'Skyss',
	position: 'top_left',
	config: {
		showPlatform: true,
		maxItems: 8,
		stops: [
			{
				stopId: "12011092"
			},
			{
				stopId: "12011022"
			},
		]
	}
},
```

# Configuration options

These are the valid configuration options you can put inside the config array above:

Configuration option | Comment | Default 
---|---|---
stops | Array of stops. See below | Empty array
maxItems | Number of journeys to display | 5 
showHeader | Set this to true to show header above the journeys | false
showStopName | Display custom stop name for each stop. You can override the name by adding `stopName` to the stop you want to override (See below) | false
maxNameLength | Some stop names can be very long and ruin the layout of the module. Set this to how many characters you max want to display.  | Not set
showPlatform | Set this to true to get the names of the platforms. Set this to true to check the name of the platform if you need to filter  | false
humanizeTimeTreshold | If time to next journey is below this value, it will be displayed as "x minutes" instead of time | 15 
serviceReloadInterval | Refresh rate in MS for how often we call Skyss's web service. NB! Don't set it too low! | 30000 
timeReloadInterval | Refresh rate how often we check if we need to update the time in the GUI | 1000 
animationSpeed | How fast the animation changes when updating mirror - in milliseconds | 0  
fade | Set this to true to fade list from light to dark | true  
fadePoint | Start fading on 1/4th of the list | 0.25

## Stops
You have to configure at least one stop. The module uses the same stop ID's used in the [avgangsvisning](https://avgangsvisning.skyss.no/) mode on skyss' website. just go to [this link](https://avgangsvisning.skyss.no/) and select the stops you want (different directions of the same stop each have their own stop). Fill out the rest of the form randomly so the website generates a url. In this url you'll find each stop as a data argument separated by commas. Example:
https://avgangsvisning.skyss.no/board/#/?stops= **12011093,12011092,12011022,12011021** &viewFreq=10000&terminal=true&colors=dark&name=name

Stop option | Comment 
---|---
stopId | Id of stop
<!-- platformFilter | The names of the platforms you want to see. Please temporarely enable `showPlatformName` in the module configuration to get the correct platform names before you configure this option. If these names aren't valid, nothing will be displayed. -->
<!-- timeToThere | How long time in minutes does it take for you to get to this stop? It's no point showing journeys that till go in 1 minute if it takes you 5 minutes to get there... -->

Example:
```
stops: [
	{
		stopId: "12011092"
	},
	{
		stopId: "12011022"
	}
]

``` 
## Translations

This modules is translated to the following languages:

Language | Responsible
---|---
nb (Norwegian) | Cato Antonsen
en (English) | Cato Antonsen

If you add other languages, please make a PR or drop me a line!

# Future enhanchements

1. ~~Generate a new auth token for each client~~
1. Show deviations
1. ~~Add filter for platforms~~ (Each platform has it's own stop now. Simply don't include it)
1. Add filter for departures that are too close to make


CHANGELOG.md
# MMM-Skyss Change Log

## [1.3.0s1.1.1] - 2018-01-18
- Temporary fix on an update issue which needed to be addressed quickly.

## [1.3.0s1.1.0] - 2018-01-18
- Changed API endpoint to giantleap.no which has a different format and changed a bunch of behavior.

## [1.3.0s1.0.0] - 2017-11-01
- Ruter app has now been forked from the ruter app. version naming is now [(original version)s(Fork version)]

## [1.3.0] - 2017-08-19
- The name of the stop will now automatically be fetched if you set `showStopName`. But you can still override it with the stop config value: `stopName`. See [documentation](README.md "MMM-Ruter Documentation") for more information.

## [1.2.0] - 2017-08-13
- Added the option to display a custom stop name in the list. New module config value: `showStopName` and new stop config value: `stopName`. See [documentation](README.md "MMM-Ruter Documentation") for more information.
- Fixed time format

## [1.1.0] - 2017-04-17

- Now it's possible to add multiple instances of the module
- Added some padding between line number and stop name
- Set default animationSpeed to 0 to prevent "blinking" when module updates often
- Moved all service calls and logic from node_helper.js ("backend") to MMM-Ruter.js ("frontent") to simplify a rather complex logic. My initial goal by putting everything in the backend, was to reduce number of calls to the service. That was not a good design decision...


## [1.0.2] - 2016-11-01

- Added timeToThere config value to stops
- Fixed duplication of journeys if manually refreshed

## [1.0.1] - 2016-10-24

- Fixed stupid this/self-bug in the refresh-code that I managed to introduce just before initial commit

## [1.0.0] - 2016-10-23

- Initial version


node_helper.js
const NodeHelper = require("node_helper");
const https = require("https");
module.exports = NodeHelper.create({
    start: function(){
        console.log("Starting module: " + this.name)
    },
    socketNotificationReceived: function (notification, payload) {
        var self = this;
        if (notification == "getstop") {
            const options = {
                hostname: 'skyss.giantleap.no',
                path: payload
            };
            var req = https.get(options, (res)=>{
                res.setEncoding('utf8');
                var data = "";
                res.on('data', (chunk) => {
                    data = data.concat(chunk);
                });

                res.on('end', ()=>{
                    self.sendSocketNotification("getstop", {response:data});
                });
            });
            req.on('error', (e) => {
                self.sendSocketNotification("getstop", {err:e});
            });
            req.end();
        }
    }
});


MMM-Skyss.js
/* Magic Mirror
 * Module: Ruter
 *
 * By Cato Antonsen (https://github.com/CatoAntonsen)
 * MIT Licensed.
 */

Module.register("MMM-Skyss",{

    // Default module config.
    defaults: {
        timeFormat: null,              // This is set automatically based on global config
        showHeader: false,             // Set this to true to show header above the journeys (default is false)
        showPlatform: false,           // Set this to true to get the names of the platforms (default is false)
        showStopName: false,           // Show the name of the stop (you have to configure 'name' for each stop)
        maxItems: 5,                   // Number of journeys to display (default is 5)
        humanizeTimeTreshold: 15,      // If time to next journey is below this value, it will be displayed as "x minutes" instead of time (default is 15 minutes)
        serviceReloadInterval: 30000,  // Refresh rate in MS for how often we call Skyss' web service. NB! Don't set it too low! (default is 30 seconds)
        animationSpeed: 0,             // How fast the animation changes when updating mirror (default is 0 second)
        fade: true,                    // Set this to true to fade list from light to dark. (default is true)
        fadePoint: 0.25                // Start on 1/4th of the list.
    },

    getStyles: function () {
        return ["skyss.css"];
    },

    getScripts: function() {
        return [];
    },

    getTranslations: function() {
        return {
            en: "translations/en.json",
            nb: "translations/nb.json"
        }
    },

    start: function() {
        console.log(this.translate("STARTINGMODULE") + ": " + this.name);

        this.journeys = [];
        this.previousJourneys = [];
        var self = this;

         // Set locale and time format based on global config
        if (config.timeFormat === 24) {
            this.config.timeFormat = "HH:mm";
        } else {
            this.config.timeFormat = "h:mm A";
        }

        // Just do an initial poll. Otherwise we have to wait for the serviceReloadInterval
        self.startPolling();

        setInterval(function() {
            self.startPolling();
        }, this.config.serviceReloadInterval);
    },

    getDom: function() {
        if (this.journeys.length > 0) {

            var table = document.createElement("table");
            table.className = "ruter small";

            if (this.config.showHeader) {
                table.appendChild(this.getTableHeaderRow());
            }

            for(var i = 0; i < this.journeys.length; i++) {

                var journey = this.journeys[i];
                var tr = this.getTableRow(journey);

                // Create fade effect. <-- stolen from default "calendar" module
                if (this.config.fade && this.config.fadePoint < 1) {
                    if (this.config.fadePoint < 0) {
                        this.config.fadePoint = 0;
                    }
                    var startingPoint = this.journeys.length * this.config.fadePoint;
                    var steps = this.journeys.length - startingPoint;
                    if (i >= startingPoint) {
                        var currentStep = i - startingPoint;
                        tr.style.opacity = 1 - (1 / steps * currentStep);
                    }
                }

                table.appendChild(tr);
            }

            return table;
        } else {
            var wrapper = document.createElement("div");
            wrapper.innerHTML = this.translate("LOADING");
            wrapper.className = "small dimmed";
            return wrapper;
        }

    },

    startPolling: function() {
        var self = this;

        var promise = new Promise((resolv) => {
            this.getStopInfo(this.config.stops, function(err, result) {
                resolv(result);
            });
        });

        promise.then(function(promiseResults) {
            if (promiseResults.length > 0) {
                var allJourneys = [];
                for(var i=0; i < promiseResults.length; i++) {
                    allJourneys = allJourneys.concat(promiseResults[i])
                }

                allJourneys.sort(function(a,b) {
                    var dateA = new Date(a.time.Timestamp);
                    var dateB = new Date(b.time.Timestamp);
                    return dateA - dateB;
                });

                self.journeys = allJourneys.slice(0, self.config.maxItems);

                self.updateDom();
            }
        });
    },

    getStopInfo: function(stopItems, callback) {
        var self = this;

        var HttpClient = function() {
            this.get = function(requestUrl, requestCallback) {
                // var httpRequest = new XMLHttpRequest();
                // httpRequest.onreadystatechange = function() {
                //     if (httpRequest.readyState == 4 && httpRequest.status == 200){
                //         requestCallback(httpRequest.responseText);
                //     }
                // };

                // httpRequest.open("GET", requestUrl, true);
                // httpRequest.setRequestHeader("Authorization", "");
                // httpRequest.send(null);
                self.requests.push(requestCallback);
                self.sendSocketNotification("getstop", requestUrl);
            }
        }

        // var shouldAddPlatform = function(platform, platformFilter) {
        //     if (platformFilter == null || platformFilter.length == 0) { return true; } // If we don't add any interesting platformFilter, then we asume we'll show all
        //     for(var i=0; i < platformFilter.length; i++) {
        //         if (platformFilter[i] === platform) { return true; }
        //     }

        //     return false;
        // };

        // var departureUrl = function() {
        //     var dateParam = "";
        //     if (stopItem.timeToThere) {
        //         var min = stopItem.timeToThere;
        //         var timeAhead = moment(moment.now()).add(min, "minute").format().substring(0, 16);
        //         console.log("Looking for journeys " + min + " minutes ahead in time.");
        //         dateParam = "?datetime=" + timeAhead;
        //     } else {
        //         console.log("Looking for current journeys");
        //     }

        //     return "http://reisapi.ruter.no/StopVisit/GetDepartures/" + stopItem.stopId + dateParam;
        // };

        var stopUrl = function() {
            return "/public/departures?Hours=12&StopIdentifiers=" + stopItems.map(stopItem => stopItem.stopId).join();
        };

        var client = new HttpClient();

        client.get(stopUrl(), function(stopResponse) {
            var departure = JSON.parse(stopResponse);
            var times = departure.PassingTimes;

            var allStopItems = [];

            for(var j = 0; j < times.length; j++) {
                var journey = times[j];
                var stop = departure.Stops[journey.StopIdentifier];
                allStopItems.push({
                    stopId: journey.StopIdentifier,
                    stopName: stop.PlaceDescription,
                    lineName: journey.RoutePublicIdentifier,
                    destinationName: journey.TripDestination,
                    service: stop.ServiceModes[0],
                    time: {
                        Timestamp: journey.AimedTime,
                        Status: journey.Status,
                    },
                    platform: journey.Platform
                });
            }
            callback(null, allStopItems);
        })
    },

    getTableHeaderRow: function() {
        var thLine = document.createElement("th");
        thLine.className = "";
        thLine.appendChild(document.createTextNode(this.translate("LINEHEADER")));

        var thDestination = document.createElement("th");
        thDestination.className = "";
        thDestination.appendChild(document.createTextNode(this.translate("DESTINATIONHEADER")));

        var thPlatform = document.createElement("th");
        thPlatform.className = "";
        thPlatform.appendChild(document.createTextNode(this.translate("PLATFORMHEADER")));

        var thStopName = document.createElement("th");
        thStopName.className = "";
        thStopName.appendChild(document.createTextNode(this.translate("STOPNAMEHEADER")));

        var thTime = document.createElement("th");
        thTime.className = "time";
        thTime.appendChild(document.createTextNode(this.translate("TIMEHEADER")));

        var thead = document.createElement("thead");
        thead.addClass = "xsmall dimmed";
        thead.appendChild(document.createElement("th"));
        thead.appendChild(thLine);
        thead.appendChild(thDestination);
        if (this.config.showStopName) { thead.appendChild(thStopName); }
        if (this.config.showPlatform) { thead.appendChild(thPlatform); }
        thead.appendChild(thTime);

        return thead;
    },

    getTableRow: function(journey) {
        var tdIcon = document.createElement("td");
        var imageFA;
        switch (journey.service) {
        case "Bus":
        case "Express":
        case "Airport bus":
            imageFA = "bus";
            break;
        case "Light rail":
            imageFA = "subway";
            break;
        case "Ferry":
        case "Boat":
            imageFA = "ship";
            break;
        case "Train":
            imageFA = "train";
            break;
        default:
            imageFA = "rocket";
            break;
        }
        tdIcon.className = "fa fa-"+imageFA;

        var tdLine = document.createElement("td");
        tdLine.className = "line";
        var txtLine = document.createTextNode(journey.lineName);
        tdLine.appendChild(txtLine);

        var tdDestination = document.createElement("td");
        tdDestination.className = "destination bright";
        tdDestination.appendChild(document.createTextNode(journey.destinationName));

        if (this.config.showPlatform) {
            var tdPlatform = document.createElement("td");
            tdPlatform.className = "platform";
            tdPlatform.appendChild(document.createTextNode(journey.platform));
        }

        if (this.config.showStopName) {
            var tdStopName = document.createElement("td");
            tdStopName.className = "light";
            tdStopName.appendChild(document.createTextNode(journey.stopName));
        }

        var tdTime = document.createElement("td");
        if (journey.time.Status != "Schedule") {
            tdTime.className = "time light sanntid";
        } else {
            tdTime.className = "time light";
        }
        tdTime.appendChild(document.createTextNode(this.formatTime(journey.time.Timestamp)));

        var tr = document.createElement("tr");
        tr.appendChild(tdIcon);
        tr.appendChild(tdLine);
        tr.appendChild(tdDestination);
        if (this.config.showStopName) { tr.appendChild(tdStopName); }
        if (this.config.showPlatform) { tr.appendChild(tdPlatform); }
        tr.appendChild(tdTime);

        return tr;
    },

    formatTime: function(t) {
        var now = new Date();
        var tti = new Date(t);
        var diff = tti - now;
        var min = Math.floor(diff/60000);

        if (min == 0) {
            return this.translate("NOW");
        } else if (min == 1) {
            return this.translate("1MIN");
        } else if (min < this.config.humanizeTimeTreshold) {
            return min + " " + this.translate("MINUTES");
        } else {
            return tti.getHours() + ":" + ("0" + tti.getMinutes()).slice(-2);
        }
    },

    socketNotificationReceived: function(notification, payload) {
        var self = this;
        Log.log(this.name + " recieved a socket notification: " + notification);
        if (notification == "getstop") {
            if (payload.err) {
                throw payload.err;
            } else {
                self.requests.shift()(payload.response);
            }
        }
    },

    requests: []
});
